### ISSUE TITLE
val_loss don't converge, but val_acc is improving after using mnist_mlp.py

### ISSUE THREAD

**Issue by GatechLW** [fchollet/keras#5359]
Dear all, 

Keras is new to me! I just play with provided example script: mnist_mlp.py (https://github.com/fchollet/keras/blob/master/examples/mnist_mlp.py)
The accuracy is very good,  but I found a interesting thing: although val_acc is improving, val_loss  don't converge, more badly, after 20 epoch, the val_loss is larger than val_loss at first epoch. I have attached my training log, please check! Could you tell me the reason why this happen?

Thanks in advance!

![log](https://cloud.githubusercontent.com/assets/25702394/22852169/609d58f4-f004-11e6-81c1-ca3a5a7d6420.png)

**Comment by joelthchao** [comment]
It's due to over-fitting. You can observe that train_loss is decreasing all the way and val_loss increases after 7 epoch, which is symptom of over-fitting.

**Comment by unrealwill** [comment]
I DON'T believe you are over-fitting.
Your validation accuracy is improving.
That is a feature of online learning. 

http://stackoverflow.com/questions/40910857/how-to-interpret-increase-in-both-loss-and-accuracy

You are doing a good job with regularization.

What happens is that you are using a log-loss, and the model is trying to reduce the average loss. There is two main way of reducing an average : either you reduce the worst example a lot (that what mean square error does), or you try to reduce everyone a little (that's what log loss does). By doing that, and ensuring that you don't have enough parameter to overfit (i.e. memorise) your training example by regularizing them enough, you are learning to generalize well, and thus your model performance on unseen data should keep improving.

I think I saw this explanation when watching one of those online-course type of thing but can't remember the exact place. And couldn't find it anymore by doing a quick search.

In any case, when in doubt on when to stop, use an heuristic like early-stopping, this way you will be able to compare your model in a more robust way.

**Comment by joelthchao** [comment]
@unrealwill the objective is cross entropy, you can not interpret over-fitting or not by accuracy.

**Comment by unrealwill** [comment]
@joelthchao I have always thought the goal for mnist was to improve the val accuracy (i.e, the ultimate goal was to predict a single digit, instead of the distribution of uncertainty upon digit).

You are indeed correct, if your goal is to decrease validation_loss (validation_cross_entropy), then you need to regularize more.

**Comment by GatechLW** [comment]
@joelthchao @unrealwill, thanks for your kindly reply!  So Under keras framework, the most frequently used reducing over-fitting effect is using dropout? Could you tell me most frequently solutions for over-fitting? 

Thanks

**Comment by GatechLW** [comment]
@unrealwill @joelthchao  thanks for your kindly reply. As unrealwill suggested, I modify the original script, like followings:

model = Sequential()
model.add(Dense(512, input_shape=(784,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(10))
model.add(Activation('softmax'))

model.summary()

model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

But I still can observe the val_loss is firstly decrease and then increase after some epoch. I can repeat @unrealwill 's reported result. 

![log](https://cloud.githubusercontent.com/assets/25702394/22854967/df290dda-f045-11e6-9041-50f5e90edb8d.png)

**Comment by GatechLW** [comment]
@unrealwill @joelthchao thanks for your kindly reply. As joelthchao suggested, I modify the original script, like followings:

model = Sequential()
model.add(Dense(512, input_shape=(784,)))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dropout(0.5))
model.add(Dense(10))
model.add(Activation('softmax'))

model.summary()

model.compile(loss='categorical_crossentropy',
optimizer='adam',
metrics=['accuracy'])

But I still can observe the val_loss is firstly decrease and then increase after some epoch. I cannot repeat @joelthchao 
![log](https://cloud.githubusercontent.com/assets/25702394/22855038/08dd638c-f047-11e6-9f9d-a878fc26dfc1.png)
 's reported result.

**Comment by GatechLW** [comment]
@unrealwill @joelthchao, One more question, after modifying dropout parameter and optimizer options, If I still can observe overfitting,  can I use call back to save best model. Namely, I just run 100 epoch, and then select smallest val_loss or biggest val_acc using following command:
ModelCheckpoint(kfold_weights_path, monitor='val_loss', save_best_only=True, verbose=0),
or 
ModelCheckpoint(kfold_weights_path, monitor='val_acc', save_best_only=True, verbose=0),

**Comment by joelthchao** [comment]
@GatechLW 
Over-fitting is inevitable once your model is still too powerful. This is a trade-off problem:
1. **Using less powerful model** and easy to prevent over-fitting, however, you might get worse performance.
2. Using powerful model and **spend more effort** on fighting over-fitting (e.g., dropout, weight decay, finetune from pre-trained model, l1 or l2 weight regularization, shared weight, adding noise and data augmentation) and get better performance.
And, yes, you can do that. Early stop is another way to prevent over-fitting.

**Comment by GatechLW** [comment]
@joelthchao, OK, your explanation is clear! thanks! One thing about ModelCheckpoint  I am still a confused with what you said. Namely I think Early stop is different with ModelCheckpoint.  ModelCheckpoint(kfold_weights_path, monitor='val_loss', save_best_only=True, verbose=0),

For example, if set your total epoch is 100. Early stop might be stopped when the loss is increased out of your tolerance. e.g., it might only execute 50 epoch and save the last epoch parameter as final neural network parameter. 

However, for ModelCheckpoint, I think it will execute 100 epochs, and then it will automatically save the parameter which give best loss during this 100 epochs.
